### ISSUE TITLE
Establish Distinction between public and private headers

### ISSUE THREAD

**Issue by JoshuaSBrown** [lanl/parthenon#239]
A clear distinction in missing between what in parthenon should be publicly exposed. This makes it easy to break down stream code that depends on parthenon as a library as internal changes are likely to break down stream packages.

**Comment by Yurlungur** [comment]
This is clearly important and something we need to eventually clarify and stick to. Thanks for bringing it up.

That said, I'm not sure we're ready to solidify this... It's true that we're beginning to build things downstream with `parthenon`, but we haven't written enough, I think, to know what should and should not be public. And also `parthenon` itself is still changing pretty rapidly. I think trying to keep up an API while parthenon is still in flux will be a challenge.

**Comment by AndrewGaspar** [comment]
I agree with @JoshuaSBrown. My concern with waiting is that it ossifies the exported APIs.

**Comment by AndrewGaspar** [comment]
My proposal for this is:
- Create separate `include` and `internal` directories for public and private headers, respectively.
- Move every currently existing header file to `include`.
- Going forward, names should be deliberately placed in either the `include` or `internal` directory
- For names that must be published but are ultimately implementation details, they should go into the `parthenon::internal` namespace

**Comment by Yurlungur** [comment]
That seems reasonable to me. As discussed with @nmsriram yesterday, I also think we should try to deliberately make better use of forward declarations if we can, to reduce the amount of private headers that end up "published" via a tree of includes.

**Comment by JoshuaSBrown** [comment]
Well, I will start this process by moving all headers to a public includes folder. Will probably wait until after this has been merged though. https://github.com/lanl/parthenon/pull/238
